cli/internal/publish/images.go

```diff
@@ -0,0 +1,367 @@
+	if err != nil {
+		return "", err
+	}
+	if len(ims) == 0 {
+		return contents, nil
+	}
+
+	fmt.Println("Preparing to upload images...")
+
+	reqs := make([]cloudquery_api.TeamImageCreate, 0, len(ims))
+	for k := range ims {
+		refParts := strings.SplitN(k, ":", 2)
+		checksum, name := refParts[0], refParts[1]
+
+		reqs = append(reqs, cloudquery_api.TeamImageCreate{
+			Name:     name,
+			Checksum: checksum,
+		})
+	}
+
+	resp, err := c.CreateTeamImagesWithResponse(ctx, teamName, cloudquery_api.CreateTeamImagesJSONRequestBody{Images: reqs})
+	if err != nil {
+		return "", fmt.Errorf("failed to prepare doc images: %w", err)
+	}
+	if resp.HTTPResponse.StatusCode > 299 {
+		return "", fmt.Errorf("failed preparing: %w", hub.ErrorFromHTTPResponse(resp.HTTPResponse, resp))
+	}
+
+	for _, item := range resp.JSON201.Items {
+		item := item
+
+		key := item.Checksum + ":" + item.Name
+		for i := range ims[key] {
+			imItem := ims[key][i]
+			imItem.url = item.URL
+			ims[key][i] = imItem
+		}
+	}
+
+	contents, err = replaceMarkdownImages(contents, ims)
+	if err != nil {
+		return "", fmt.Errorf("failed replacing markdown: %w", err)
+	}
+
+	fmt.Println("Uploading images...")
+	eg, egCtx := errgroup.WithContext(ctx)
+	eg.SetLimit(4)
+
+	// Iterate again to upload
+	for _, item := range resp.JSON201.Items {
+		if item.UploadURL == nil {
+			// Already exists in bucket
+			continue
+		}
+		item := item
+		fileref := ims[item.Checksum+":"+item.Name][0].absFile
+		eg.Go(func() error {
+			return uploadImage(egCtx, *item.UploadURL, fileref)
+		})
+	}
+
+	if err := eg.Wait(); err != nil {
+		return "", fmt.Errorf("failed to upload doc images: %w", err)
+	}
+
+	return contents, nil
+}
+
+func findMarkdownImages(contents, docDir string) (map[string][]imageReference, error) {
+	imf := &imageFinder{
+		docDir: docDir,
+	}
+	p := goldmark.New(
+		goldmark.WithParserOptions(
+			parser.WithASTTransformers(util.Prioritized(imf, 999999)),
+		),
+	)
+	if err := p.Convert([]byte(contents), io.Discard); err != nil {
+		return nil, fmt.Errorf("failed to parse markdown: %w", err)
+	}
+	if imf.err != nil {
+		return nil, imf.err
+	}
+
+	return imf.images, nil
+}
+
+func replaceMarkdownImages(contents string, ims map[string][]imageReference) (string, error) {
+	reps := make([]imageReference, 0, len(ims))
+	test := bytes.Repeat([]byte("."), len(contents))
+	for _, imList := range ims {
+		for _, ir := range imList {
+			if ir.startPos == 0 && ir.endPos == 0 {
+				// return "", fmt.Errorf("unknown range for image %q", ir.ref)
+				continue // skip
+			}
+			if bytes.ContainsAny(test[ir.startPos:ir.endPos], "!") {
+				return "", fmt.Errorf("found overlapping range: %d-%d", ir.startPos, ir.endPos)
+			}
+			// test = append(append(test[:ir.startPos], bytes.Repeat([]byte("!"), ir.endPos-ir.startPos)...), test[ir.endPos:]...)
+			for i := ir.startPos; i < ir.endPos; i++ {
+				test[i] = '!'
+			}
+
+			ir := ir
+			reps = append(reps, ir)
+		}
+	}
```

Comment:
I think this could be simplified / improved by using sweep-line approach, at least in terms of memory & efficiency:

 1. create one list of all important positions (start / end) `struct {pos int, start bool}`, sort it by position. while doing this, check that start < end in all cases (error out if not)
 2. keep track of number of open ranges (`open := 0`)
 3. iterate over list; if start node then `open++`, if end node `open--`. If open is > 1, we have overlapping ranges and should error out.

Reply:
Done in [`3e8252a` (#16033)](https://github.com/cloudquery/cloudquery/pull/16033/commits/3e8252a5c51e0fd87305a3cb4a2c36ddcfaeda4b)